// ngx-float-ui/esm2022/lib/directives/ngx-float-ui/ngx-float-ui-loose.directive.mjs
import { Directive, Inject, Input } from "@angular/core";
import { NGX_FLOAT_UI_DEFAULTS } from "../../models/ngx-float-ui-defaults.model";
import { NgxFloatUiDirective } from "./ngx-float-ui.directive";
import * as i0 from "@angular/core";
export class NgxFloatUiLooseDirective extends NgxFloatUiDirective {
    set floatUiLoose(newValue) {
        this.floatUi = newValue;
    }
    set loosePlacement(newValue) {
        this.placement = newValue;
    }
    set looseTrigger(newValue) {
        this.showTrigger = newValue;
    }
    constructor(changeDetectorRef, elementRef, vcr, popperDefaults = {}) {
        super(changeDetectorRef, elementRef, vcr, popperDefaults);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiLooseDirective, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: NGX_FLOAT_UI_DEFAULTS }], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.4", type: NgxFloatUiLooseDirective, isStandalone: true, selector: "[floatUiLoose]", inputs: { floatUiLoose: "floatUiLoose", loosePlacement: "loosePlacement", looseTrigger: "looseTrigger" }, exportAs: ["floatUiLoose"], usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiLooseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[floatUiLoose]",
                    exportAs: "floatUiLoose",
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGX_FLOAT_UI_DEFAULTS]
                }] }], propDecorators: { floatUiLoose: [{
                type: Input
            }], loosePlacement: [{
                type: Input
            }], looseTrigger: [{
                type: Input
            }] } });
