// ngx-float-ui/esm2022/lib/ngx-float-ui.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
//
import { NgxFloatUiDirective } from "./directives/ngx-float-ui/ngx-float-ui.directive";
import { NgxFloatUiLooseDirective } from "./directives/ngx-float-ui/ngx-float-ui-loose.directive";
import { NgxFloatUiContentComponent } from "./components/ngx-float-ui-content/ngx-float-ui-content.component";
import { NGX_FLOAT_UI_DEFAULTS } from "./models/ngx-float-ui-defaults.model";
import * as i0 from "@angular/core";
export function provideNgxFloatUiOptions(config = {}) {
    return [
        { provide: NGX_FLOAT_UI_DEFAULTS, useValue: config },
    ];
}
export class NgxFloatUiModule {
    static forRoot(popperBaseOptions) {
        return {
            ngModule: NgxFloatUiModule,
            providers: [
                provideNgxFloatUiOptions(popperBaseOptions)
            ]
        };
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    /** @nocollapse */ static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiModule, imports: [CommonModule,
            NgxFloatUiContentComponent,
            NgxFloatUiDirective,
            NgxFloatUiLooseDirective], exports: [NgxFloatUiContentComponent,
            NgxFloatUiDirective,
            NgxFloatUiLooseDirective] });
    /** @nocollapse */ static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiModule, providers: [
            provideNgxFloatUiOptions()
        ], imports: [CommonModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NgxFloatUiModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        NgxFloatUiContentComponent,
                        NgxFloatUiDirective,
                        NgxFloatUiLooseDirective
                    ],
                    exports: [
                        NgxFloatUiContentComponent,
                        NgxFloatUiDirective,
                        NgxFloatUiLooseDirective
                    ],
                    providers: [
                        provideNgxFloatUiOptions()
                    ]
                }]
        }] });
